components/selection/selection-observer-mixin.js

```diff
@@ -33,8 +41,31 @@ export const SelectionObserverMixin = superclass => class extends superclass {
 
 	disconnectedCallback() {
 		super.disconnectedCallback();
-		if (!this._provider) return;
-		this._provider.unsubscribeObserver(this);
+		if (this._provider) this._provider.unsubscribeObserver(this);
 	}
 
+	updated(changedProperties) {
+		super.updated(changedProperties);
+
+		if (!changedProperties.has('selectionFor')) return;
+
+		if (this._selectionForObserver) this._selectionForObserver.disconnect();
+		if (this._provider) this._provider.unsubscribeObserver(this);
+
+		this._provider = this.getRootNode().querySelector(`#${cssEscape(this.selectionFor)}`);
+		if (this._provider) this._provider.subscribeObserver(this);
+
```

Comment:
GitHub lost my comment that I put in here from yesterday, so I'll try to re-create it...

I think we now need to handle the case where the provider is removed from the DOM but the observer is not. I think that would be easiest to do from the SelectionMixin's `disconnectedCallback`? That way if _both_ the provider and observer are removed, they still get detached.

And then once that's happening, I'm wondering if we could refactor this observer code... does it always run at least once? If so, I think you could just let the observer code do the initial fetching of the provider. If not, then can we avoid wiring it up at all if the provider is found right away?

Lastly, the observer should be disconnected in this component's `disconnectedCallback`.

Reply:
> I think we now need to handle the case where the provider is removed from the DOM but the observer is not. I think that would be easiest to do from the SelectionMixin's `disconnectedCallback`? That way if both the provider and observer are removed, they still get detached.

So that is actually handled, because the mutation observer runs whenever any node is added or deleted.  So if the provider is deleted, the code runs and [checks that it can still find it's provider](const selectionComponent = this.getRootNode().querySelector(`#${cssEscape(this.selectionFor)}`);), and if not it'll [set the its provider to null](https://github.com/BrightspaceUI/core/pull/1549/files#diff-a7412e30193c0f86945842eaebc7ea7709b8f1b5a541f5b42a7bd56026f5735aR61).  And then if the provider is added back, the same code will run and add it back.  I definitely missed adding a test for that case though, so I'll add that now!

I originally _did_ have the provider's `disconnectedCallback` setting all its subscribers' `_provider`s to null, but it was overkill when the `mutationObserver` code was also handling that case.  But we could change it to more like what you're saying - try to get the provider in `updated` when `selectionFor` is set or changed.  If we can, all good - don't setup a `mutationObserver` until our provider tells us it's going away in its `disconnectedCallback`.  If we can't find it, setup a `mutationObserver` to keep checking every time a node is added or deleted until we get one.

I liked that this way didn't involve the `SelectionMixin` having to care at all, but the way above does seem like it'd be more performant with less `mutationObserver`s running

Reply:
Apologies I missed how that would set things to `null` -- I'm good with how you have it set up. 👍 

That last question remains: does the mutation observer always fire at least once when you set it up, or if it finds the provider right away does it never fire? Just wondering if you could remove the duplicate 2 lines that query for the provider and subscribe.

Reply:
> does the mutation observer always fire at least once when you set it up

From what I can tell, it does not - it will only fire for added/removed nodes after `observe` is called.  Meaning it never fires for the demo page, and only fires in specific situations on the filter page (like loading, where the list pops in later).  I added a bit of code to those two lines to handle clearing the selection info though, so I wrapped it in a function.

Reply:
> Lastly, the observer should be disconnected in this component's disconnectedCallback.

Done, and I also added it to the `LabelledMixin` I was using as an example

Reply:
Cool that addresses my code duplication concern regardless.